### ISSUE TITLE
Slow initial startup show ES title screen for >3 minutes with no progress bar.

### ISSUE THREAD

**Issue by rsn8887** [RetroPie/EmulationStation#457]
I have many thousands of roms present, some of them without game list files. In this situation, ES takes forever to load on power up (~3 minutes). I think it is looping over all the roms to find the filenames. I don't understand why this is so slow, since it is not scraping the roms.

It would be nice to a) speed up the launch of ES on boot, b) display some kind of progress or at least a number in the bottom left corner, such as "6,125 roms found" with a number that increases.

**Comment by billywade** [comment]
From what I can tell, there were plans to fix this in upstream. Whether or not that's happening on the fork, no idea.

https://github.com/Aloshi/EmulationStation/issues/191

**Comment by aarononeal** [comment]
I added a comment in the upstream describing the problem. It's not XML vs. SQLite.

**Comment by pjft** [comment]
@aarononeal upstream isn't maintained anymore, as far as I'm aware, but if you'd want to submit a PR here we'll gladly review it. Thanks!

**Comment by rsn8887** [comment]
The proposed fix mentions that the user should set "Parse Gamelists Only" to true. If I do that, my boot time is already very short. My problem ist that I don't want to set "Parse Gamelists Only". I want Emulationstation to find and display new roms, and to display sub folders correctly. For that reason, I set "Parse Gamelists Only" to false.

I still don't understand why parse gamelists only should be true. Again: It doesn't have to unzip or scrape roms, it just needs to go through the directories to find new rom files. That should take seconds, not minutes, even with 6000 roms.

**Comment by aarononeal** [comment]
> Of course it is short, because Emulationstation just parses a few XML files then.

That's not true. Without this fix, it checks the filesystem for every entry in the XML. This is extremely slow over high latency networks or media.

**Comment by rsn8887** [comment]
Ok maybe it is slow for you, but for me "Parse Gamelists Only" was always quite fast. Maybe 10-20 seconds boot time. But not minutes. The problem was when that setting was OFF.

**Comment by aarononeal** [comment]
You didn't mention the media you're loading from.

**Comment by rsn8887** [comment]
USB Stick.

**Comment by rsn8887** [comment]
The problem with the setting being ON is that only correctly scraped roms will be displayed. For example the Amiga emulator will not display any of my config files etc if that setting is ON, because well the scraper doesn't scrape them. I would still like to see them without editing xml files by hand.

**Comment by rsn8887** [comment]
Who is the current maintainer of Emulationstation by the way? The original repo seems to be dead with 425 issues open.

**Comment by aarononeal** [comment]
There are 2 main paths, one parsing only game lists only and one scanning folders.

I optimized the parse only path by eliminating all `stat` calls to the filesystem. In my high latency test case, this had huge (160x faster) impact.

Looking at the other code path of scanning folders, I see `stat` is called for every entry in the gamelist *multiple* times. Depending on the media that's not as cheap as you might think. In some cases it results in updating last access time for each file. This may be contributing to the performance problems in that path as well.

**Comment by hex007** [comment]
I have some time to spare this week. Will take a look, unless any other maintainer is interested.

**Comment by hex007** [comment]
`lstat` is used mostly to follow symblinks for files. I am working on 2 approaches for this, either completely disable the symblink resolution flow using a param in settings or hard code the symblink resolution and work with the OS to auto follow the links. The second seems to be harder to figure out. Might take while so will keep posting status on this thread.
